torch/optim/optimizer.py

```diff
@@ -294,25 +381,33 @@ def wrapper(*args, **kwargs):
         return wrapper
 
     @staticmethod
-    def _group_tensors_by_device_and_dtype(tensorlistlist, with_indices=False):
+    def _group_tensors_by_device_and_dtype(
+        tensorlistlist: List[List[torch.Tensor]],
+        with_indices: bool = False,
+    ) -> Union[
+        Dict[Tuple[None, None], List[Union[List[torch.Tensor], List[int]]]],
+        Dict[Tuple[torch.device, torch.dtype], List[Union[List[torch.Tensor], List[int]]]],
+    ]:
         """Groups a list of lists of tensors by device and dtype.
         Skips this step if we are compiling since this will occur during inductor lowering."""
         if is_compiling():
-            if with_indices:
```

Comment:
I'd prefer the old if statement instead of the ternary op because in the old implementation, indices doesn't get materialized if with_indices is false. There is a perf implication here.